anvil/src/udev.rs

```diff
@@ -733,30 +777,71 @@ impl AnvilState<UdevData> {
-            .ok()
-            .and_then(|x| x.try_get_render_node().ok().flatten())
-            .unwrap_or(node);
+        let mut try_initialize_gpu = || {
+            let display = unsafe { EGLDisplay::new(gbm.clone()).map_err(DeviceAddError::AddNode)? };
+            let egl_device = EGLDevice::device_for_display(&display).map_err(DeviceAddError::AddNode)?;
 
-        self.backend_data
-            .gpus
-            .as_mut()
-            .add_node(render_node, gbm.clone())
-            .map_err(DeviceAddError::AddNode)?;
+            let render_node = egl_device
+                .try_get_render_node()
+                .map_err(DeviceAddError::AddNode)?
+                .ok_or(DeviceAddError::NoRenderNode)?;
+            self.backend_data
+                .gpus
+                .as_mut()
+                .add_node(render_node, gbm.clone())
+                .map_err(DeviceAddError::AddNode)?;
+
+            std::result::Result::<DrmNode, DeviceAddError>::Ok(render_node)
+        };
+
+        let render_node = try_initialize_gpu()
+            .inspect_err(|err| {
+                warn!(?err, "failed to initialize gpu");
+            })
+            .ok();
+
+        let allocator = render_node
+            .is_some()
+            .then(|| GbmAllocator::new(gbm.clone(), GbmBufferFlags::RENDERING | GbmBufferFlags::SCANOUT))
+            .or_else(|| {
+                self.backend_data
+                    .backends
+                    .get(&self.backend_data.primary_gpu)
+                    .or_else(|| {
+                        self.backend_data
+                            .backends
+                            .values()
+                            .find(|backend| backend.render_node == Some(self.backend_data.primary_gpu))
+                    })
+                    .map(|backend| backend.drm_output_manager.allocator().clone())
+            })
+            .ok_or(DeviceAddError::PrimaryGpuMissing)?;
+
+        let framebuffer_exporter = GbmFramebufferExporter::new(gbm.clone());
 
-        let allocator = GbmAllocator::new(gbm.clone(), GbmBufferFlags::RENDERING | GbmBufferFlags::SCANOUT);
         let color_formats = if std::env::var("ANVIL_DISABLE_10BIT").is_ok() {
             SUPPORTED_FORMATS_8BIT_ONLY
         } else {
             SUPPORTED_FORMATS
         };
-        let mut renderer = self.backend_data.gpus.single_renderer(&render_node).unwrap();
-        let render_formats = renderer.as_mut().egl_context().dmabuf_render_formats().clone();
+        let mut renderer = self
+            .backend_data
+            .gpus
+            .single_renderer(&render_node.unwrap_or(self.backend_data.primary_gpu))
+            .unwrap();
+        let render_formats = renderer
+            .as_mut()
+            .egl_context()
+            .dmabuf_render_formats()
+            .iter()
+            .filter(|format| render_node.is_some() || format.modifier == Modifier::Linear)
```

Comment:
I see, here we do the same for the global feedback. why?

Reply:
I guess passing this set to `DrmCompositor` is intentional though, right?

Reply:
right, left this in as this only changed the formats for `DrmCompositor`